File: medium/ContainerWithMostWater.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int maxArea(vector<int> &height) 
{
    int n = height.size();
    int i = 0, j = n-1;
    int max_water = 0;
    while (i < j) 
    {
        // if a[i] < a[j], then if we minus j, the result couldn't be better
        if (height[i] < height[j]) 
            max_water = max(max_water, height[i] * (j-i)), i ++;
        else 
            max_water = max(max_water, height[j] * (j-i)), j --;
    }
    
    return max_water;
}

int main() 
{
    vector<int> height;
    height.push_back(1);
    height.push_back(2);
    height.push_back(4);
    height.push_back(3);

    cout << maxArea(height) << endl;

    return 0;
}

File: medium/ContainerWithMostWater.py

class Solution:
    def maxArea(self, height):
        len_of_height = len(height)
        if len_of_height <= 0:
            return 0
        max_of_all, l_ptr, r_ptr = 0, 0, len_of_height-1
        while l_ptr != r_ptr:
            max_of_all = max(max_of_all,
                             min(height[l_ptr], height[r_ptr])
                             * (r_ptr-l_ptr))
            if height[l_ptr] >= height[r_ptr]:
                r_ptr -= 1
            else:
                l_ptr += 1
        return max_of_all

sol = Solution()
print sol.maxArea([1,1])
print sol.maxArea([3,5,2,1,3])

